ddtrace/writer.py

```diff
@@ -169,64 +218,117 @@ def _log_error_status(self, response, response_name):
                 response.msg,
             )
 
-    def _apply_filters(self, traces):
-        """
-        Here we make each trace go through the filters configured in the
-        tracer. There is no need for a lock since the traces are owned by the
-        AsyncWorker at that point.
-        """
-        if self._filters is not None:
-            filtered_traces = []
-            for trace in traces:
-                for filtr in self._filters:
-                    trace = filtr.process_trace(trace)
-                    if trace is None:
-                        break
-                if trace is not None:
-                    filtered_traces.append(trace)
-            return filtered_traces
-        return traces
-
 
-class Q(object):
+class Q(Queue, object):
     """
-    Q is a threadsafe queue that let's you pop everything at once and
-    will randomly overwrite elements when it's over the max size.
+    Q is a custom subclass of ``queue.Queue`` that:
+
+      - Adds `.close()` and `.closed()` to support closing the queue
+      - Support non-blocking, and non-error popping (return ``None`` instead of an error)
+      - Support non-blocking, and non-error pushing (overwrite a random item in the queue when max size is reached)
+      - Add `.wait()` helper to supporting sleeping until a new item is available to pop
+      - Modify `.join()` to wait until all items are removed from the queue instead of all tasks being marked as done
     """
-    def __init__(self, max_size=1000):
-        self._things = []
-        self._lock = threading.Lock()
-        self._max_size = max_size
-        self._closed = False
+    def __init__(self, maxsize=1000):
+        """
+        Constructor for Q
 
-    def size(self):
-        with self._lock:
-            return len(self._things)
+        :param maxsize: The max number of items to put into the queue
+        :type maxsize: int
+        """
+        super(Q, self).__init__(maxsize=maxsize)
+        self._closed = False
 
     def close(self):
-        with self._lock:
+        """
```

Comment:
You'll see while rebasing, but the `close` attribute is actually useless the way the Q is used.